### PR TITLE
feat(asm/pylons): detect attack on Pylons request body

### PR DESCRIPTION
**Pull request by juanjux**
## Description
Detects attacks on request body when running inside the Pylons framework.

## Checklist
- [X] Title must conform to [conventional commit](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional).
- [X] Add additional sections for `feat` and `fix` pull requests.
- [X] Ensure tests are passing for affected code.
- [ ] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Reviewer Checklist
- [ ] Title is accurate.
- [ ] Description motivates each change.
- [ ] No unnecessary changes were introduced in this PR.
- [ ] PR cannot be broken up into smaller PRs.
- [ ] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.
- [ ] Add to milestone.

### CODE REVIEW

File: ddtrace/contrib/pylons/middleware.py [hunk 1]
```diff
@@ -47,6 +51,20 @@ def _distributed_tracing(self):
     def _distributed_tracing(self, distributed_tracing):
         ddconfig.pylons["distributed_tracing"] = asbool(distributed_tracing)
 
+    def _unlist_multidict_params(self, multidict):  # type: (webob.multidict.MultiDict) -> Dict[str, Any]
+        # Pylons multidict always return values in a list. This will convert lists with a
+        # single value
+        # into the un-listed value
+        new_dict = {}  # type: Dict[str, Any]
+
+        for k, v in six.iteritems(multidict):
```
Comment (gnufede) [inline ddtrace/contrib/pylons/middleware.py:60]: nit: Maybe use `internal.compat.iteritems` instead?

File: ddtrace/contrib/pylons/middleware.py [hunk 2]
```diff
@@ -47,6 +51,20 @@ def _distributed_tracing(self):
...
+            if isinstance(v, typing.List) and len(v) == 1:
```
Comment (gnufede) [inline ddtrace/contrib/pylons/middleware.py:61]: nit: `isinstance(v, list)` so you can avoid ` import typing`

File: ddtrace/contrib/pylons/middleware.py [hunk 3]
```diff
@@ -67,12 +85,34 @@ def __call__(self, environ, start_response):
             # set analytics sample rate with global config enabled
             span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, ddconfig.pylons.get_analytics_sample_rate(use_global_config=True))
 
+            req_body = None
+
+            # XXX check how to use appsec_enabled here
+            # if ddconfig.pylons._appsec_enabled and request.method in _BODY_METHODS:
+            if request.method in _BODY_METHODS:
+                content_type = (
+                    request.content_type if hasattr(request, "content_type") else request.META["CONTENT_TYPE"]
```
Comment (gnufede) [inline ddtrace/contrib/pylons/middleware.py:94]: nit: How about `getattr(request, "content_type", request.META["CONTENT_TYPE")`
[reference](https://docs.python.org/2/library/functions.html#getattr)

File: ddtrace/contrib/pylons/middleware.py (hunk 1 above)
Comment (juanjux) [reply ddtrace/contrib/pylons/middleware.py]: Done, thanks!

File: ddtrace/contrib/pylons/middleware.py (hunk 3 above)
Comment (juanjux) [reply ddtrace/contrib/pylons/middleware.py]: Done, thanks!

File: ddtrace/contrib/pylons/middleware.py (hunk 2 above)
Comment (juanjux) [reply ddtrace/contrib/pylons/middleware.py]: Done, thanks!

**Review by gnufede — DISMISSED** [review]

**Comment by juanjux** [comment]
Thanks @avara1986 for fixing CI!

**Review by avara1986 — APPROVED** [review]

**Review by avara1986 — APPROVED** [review]

**Review by gnufede — APPROVED** [review]

**Pull request merged**